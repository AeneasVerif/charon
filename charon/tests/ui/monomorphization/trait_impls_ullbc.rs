//@ charon-args=--monomorphize --ullbc --print-ullbc --no-serialize
//@ charon-args=--start-from=crate::main
// Ensures monomorphization happens when trait implementations are involved.

use std::mem;

fn do_test<T: std::cmp::Eq>(init: T, expected: T) {
    assert!(expected == init);
}

fn main() {
    do_test::<bool>(true, true);
}
