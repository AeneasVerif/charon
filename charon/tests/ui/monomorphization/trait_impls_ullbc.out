# Final ULLBC before serialization:

fn test_crate::do_test<T>(@1: bool, @2: bool)
where
    [@TraitClause0]: core::marker::Sized<bool>,
    [@TraitClause1]: core::cmp::Eq<bool>,
{
    let @0: (); // return
    let init@1: bool; // arg #1
    let expected@2: bool; // arg #2
    let @3: (); // anonymous local
    let @4: bool; // anonymous local
    let @5: &'_ (bool); // anonymous local
    let @6: &'_ (bool); // anonymous local

    bb0: {
        storage_live(@3);
        storage_live(@4);
        storage_live(@5);
        @5 := &expected@2;
        storage_live(@6);
        @6 := &init@1;
        @4 := core::cmp::impls::{impl core::cmp::Eq for bool}#38::parent_clause0::eq<'_, '_, bool, bool>(move (@5), move (@6)) -> bb1 (unwind: bb2);
    }

    bb1: {
        if move (@4) -> bb3 else -> bb4;
    }

    bb2: {
        drop expected@2;
        drop init@1;
        unwind_continue;
    }

    bb3: {
        storage_dead(@6);
        storage_dead(@5);
        storage_dead(@4);
        storage_dead(@3);
        @0 := ();
        drop expected@2;
        drop init@1;
        @0 := ();
        return;
    }

    bb4: {
        storage_dead(@6);
        storage_dead(@5);
        panic(core::panicking::panic);
    }
}

#[lang_item("sized")]
pub trait core::marker::Sized<Self>

#[lang_item("eq")]
pub trait core::cmp::PartialEq<Self, Rhs>
{
    fn eq<'_0, '_1> = core::cmp::PartialEq::eq<'_0_0, '_0_1, Self, Rhs>
}

#[lang_item("Eq")]
pub trait core::cmp::Eq<Self>
{
    parent_clause0 : [@TraitClause0]: core::cmp::PartialEq<Self, Self>
}

impl core::cmp::impls::{impl core::cmp::PartialEq<bool> for bool}#19 : core::cmp::PartialEq<bool, bool>
{
    fn eq<'_0, '_1> = core::cmp::impls::{impl core::cmp::PartialEq<bool> for bool}#19::eq<'_0_0, '_0_1>
}

impl core::cmp::impls::{impl core::cmp::Eq for bool}#38 : core::cmp::Eq<bool>
{
    parent_clause0 = core::cmp::impls::{impl core::cmp::PartialEq<bool> for bool}#19
}

fn test_crate::main()
{
    let @0: (); // return
    let @1: (); // anonymous local

    bb0: {
        storage_live(@1);
        @1 := @Fun7<bool>[core::marker::Sized<bool>, core::cmp::impls::{impl core::cmp::Eq for bool}#38](const (true), const (true)) -> bb1 (unwind: bb2);
    }

    bb1: {
        storage_dead(@1);
        @0 := ();
        @0 := ();
        return;
    }

    bb2: {
        unwind_continue;
    }
}



